### PR TITLE
infoschema: change data structure for lookup table by ID.

### PR DESCRIPTION
**Pull request by coocood**
Use 512 of table slice to lookup table by ID.
When schema is updated, only 1/512 of memory need to be allocated.

### CODE REVIEW

File: infoschema/builder.go [hunk 1]
```diff
@@ -136,58 +158,57 @@ func (b *Builder) applyCreateTable(m *meta.Meta, roDBInfo *model.DBInfo, tableID
...
-	b.copyTables(oldIS)
+	b.copySchemasMap(oldIS)
+	copy(b.is.sortedTablesBuckets, oldIS.sortedTablesBuckets)
 	return b
 }
 
-func (b *Builder) copySchemaNames(oldIS *infoSchema) {
-	for k, v := range oldIS.schemaNameToID {
-		b.is.schemaNameToID[k] = v
+func (b *Builder) copySchemasMap(oldIS *infoSchema) {
```
Comment (ngaut) [inline infoschema/builder.go]: s/schemasMap/schemaMap/

File: infoschema/builder.go [hunk 2]
```diff
@@ -199,47 +220,60 @@ func (b *Builder) InitWithDBInfos(dbInfos []*model.DBInfo, schemaVersion int64)
 	info := b.is
 	info.schemaMetaVersion = schemaVersion
 	for _, di := range dbInfos {
-		info.schemas[di.ID] = di
-		info.schemaNameToID[di.Name.L] = di.ID
+		schTbls := &schemaTables{
+			dbInfo: di,
+			tables: make(map[string]table.Table),
```
Comment (ngaut) [inline infoschema/builder.go]: Make with length of di.Tables

File: infoschema/builder.go [hunk 3]
```diff
@@ -199,47 +220,60 @@ func (b *Builder) InitWithDBInfos(dbInfos []*model.DBInfo, schemaVersion int64)
...
 	return b, nil
 }
 
 func (b *Builder) initMemorySchemas() error {
 	info := b.is
-	info.schemaNameToID[infoSchemaDB.Name.L] = infoSchemaDB.ID
-	info.schemas[infoSchemaDB.ID] = infoSchemaDB
+	infoSchemaTblNames := &schemaTables{
+		dbInfo: infoSchemaDB,
+		tables: make(map[string]table.Table),
```
Comment (ngaut) [inline infoschema/builder.go]: Ditto.

File: infoschema/builder.go [hunk 4]
```diff
@@ -199,47 +220,60 @@ func (b *Builder) InitWithDBInfos(dbInfos []*model.DBInfo, schemaVersion int64)
...
+		bucketIdx := t.ID % bucketCount
+		info.sortedTablesBuckets[bucketIdx] = append(info.sortedTablesBuckets[bucketIdx], tbl)
 	}
 
 	perfHandle := b.handle.memSchema.perfHandle
-	psDB := perfHandle.GetDBMeta()
+	perfSchemaDB := perfHandle.GetDBMeta()
+	perfSchemaTblNames := &schemaTables{
+		dbInfo: perfSchemaDB,
+		tables: make(map[string]table.Table),
```
Comment (ngaut) [inline infoschema/builder.go]: Ditto.

File: infoschema/builder.go [hunk 5]
```diff
@@ -74,22 +79,41 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) error {
 	return nil
 }
 
+// CopySortedTables copies sortedTables for old table and new table for later modification.
+func (b *Builder) copySortedTables(oldTableID, newTableID int64) {
+	buckets := b.is.sortedTablesBuckets
+	if oldTableID != 0 {
```
Comment (ngaut) [inline infoschema/builder.go]: Please define a constant as invalid table id.

File: infoschema/builder.go [hunk 6]
```diff
@@ -74,22 +79,41 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) error {
...
+		newSortedTables := make(sortedTables, len(oldSortedTables), len(oldSortedTables)+1)
+		copy(newSortedTables, oldSortedTables)
+		buckets[newTableID%bucketCount] = newSortedTables
+	}
+}
+
 // updateDBInfo clones a new DBInfo from old DBInfo, and update on the new one.
 func (b *Builder) updateDBInfo(roDBInfo *model.DBInfo, oldTableID, newTableID int64) {
 	newDbInfo := new(model.DBInfo)
 	*newDbInfo = *roDBInfo
```
Comment (ngaut) [inline infoschema/builder.go]: newDbInfo := *roDBInfo

File: infoschema/infoschema.go [hunk 7]
```diff
@@ -81,128 +82,161 @@ const (
...
+	dbInfo := &model.DBInfo{ID: 0, Name: model.NewCIStr("test"), Tables: tbList}
+	tableNames := &schemaTables{
+		dbInfo: dbInfo,
+		tables: make(map[string]table.Table),
+	}
+	result.schemaMap["test"] = tableNames
+	for _, tb := range tbList {
+		tbl := table.MockTableFromMeta(tb)
+		tableNames.tables[tb.Name.L] = tbl
+		bucketIdx := tb.ID % bucketCount
```
Comment (ngaut) [inline infoschema/infoschema.go]: We should extract tb.ID % bucketCount to a hash function.

File: infoschema/builder.go [hunk 8]
```diff
@@ -74,22 +81,40 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) error {
 	return nil
 }
 
+// CopySortedTables copies sortedTables for old table and new table for later modification.
+func (b *Builder) copySortedTables(oldTableID, newTableID int64) {
+	buckets := b.is.sortedTablesBuckets
+	if oldTableID != invalidTableID {
```
Comment (ngaut) [inline infoschema/builder.go]: We should extract validation logic to a function.

**Comment by ngaut** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**